### DISCUSSION TITLE
How to use Quarkus Hibernate Reactive Panache with QuarkusWebSocket?

### DISCUSSION THREAD

**Discussion by SetoKaiba** [quarkusio/quarkus#32336] [Q&A]
How to use Quarkus Hibernate Reactive Panach with QuarkusWebSocket? I try to do some Panache Reative stuff. And it reports no current Vertx context found.
```java
    public Future<Void> sendInventoryList() {
        if (user == null) return null;
        Panache.withSession(() -> Mutiny.fetch(user.itemInstances).onItem().invoke(result -> {
            System.out.println(user.itemInstances.size());
            Live.ServerRefreshInventoryList.Builder builder = Live.ServerRefreshInventoryList.newBuilder();
//        builder.addInventories();
        })).subscribe().with(r -> {

        });
        return null;
    }
```
```
2023-04-02 17:50:02,112 ERROR [net.kai.uti.SharedLogger] (vert.x-eventloop-thread-3) onError: io.undertow.websockets.UndertowSession@3173d1a6, 30747869312974848: java.lang.IllegalStateException: No current Vertx context found
	at io.quarkus.hibernate.reactive.panache.common.runtime.SessionOperations.vertxContext(SessionOperations.java:186)
	at io.quarkus.hibernate.reactive.panache.common.runtime.SessionOperations.withSession(SessionOperations.java:112)
	at io.quarkus.hibernate.reactive.panache.Panache.withSession(Panache.java:29)
	at net.kaiba.live.model.runtime.LiveUser.sendInventoryList(LiveUser.java:264)
	at net.kaiba.live.model.runtime.LiveUser.init(LiveUser.java:173)
	at net.kaiba.live.websocket.LiveServerEndpoint.onMessage(LiveServerEndpoint.java:125)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

[a comment by an automated account is not shown]

**Comment by loicmathieu** [comment]
Can you give us more context ? We will need the enclosing class of the `sendInventoryList` to understand you use case.

But you can already update the code to return the result of `Panache.withSession`, this is a `Uni` which must be subscribed by the Quarkus HTTP engine not by your code.

**Comment by SetoKaiba** [comment]
@loicmathieu This is called inside the WebSocket onMessage. I need to access or persist data in Hibernate Reactive.
And I subscirbe this myself in the WebSocket onOpen. It doesn't tell me no current context. But it tell me this in onMessage method.
The quarkus-websockets is not based on vertx.

I think it's a common use that read/write database in WebSocket onOpen & onMessage method. The following example is simple enough  I think.

```java
    // This runs without problem
    @OnOpen
    public void onOpen(Session session, @PathParam("roomId") Long roomId) {
        Panache.withSession(() -> User.listAll().onItem().invoke(r -> {
            System.out.println(r.size());
        })).subscribe().with(r -> {
        });
    }

    // This runs with problem
    @OnMessage
    public void onMessage(Session session, ByteBuffer message, @PathParam("roomId") String roomId) {
        Panache.withSession(() -> User.listAll().onItem().invoke(r -> {
            System.out.println(r.size());
        })).subscribe().with(r -> {
        });
    }
```

**Reply by cescoffier** [reply]
Good question. 
I believe we should make sure that onMessage is called on a duplicated context. 

For the time being, you can use runSubscription and create an executor that will create a new duplicated context for every message. We have a Contexts utility class with helper methods.

**Reply by SetoKaiba** [reply]
> Good question. I believe we should make sure that onMessage is called on a duplicated context.
> 
> For the time being, you can use runSubscription and create an executor that will create a new duplicated context for every message. We have a Contexts utility class with helper methods.

```
        Context context = VertxContext.getOrCreateDuplicatedContext(vertx);
        VertxContextSafetyToggle.setContextSafe(context, Boolean.TRUE);
        context.runOnContext(x -> {
            Long roomId = KeygenUtil.decode(roomId36);
            logger.infof("onMessage: %s, %s, %s", session, message, roomId);
            liveManager.onMessage(session, message);
        });
```
@cescoffier This is what I'm doing. It seems that it can be run in a duplicated context.
Do you mean this?
```
//  create an executor that will create a new duplicated context for every message.
uni.runSubscriptionOn(executor);
```
![image](https://user-images.githubusercontent.com/159540/231912143-a4d9341a-714b-4f2a-8b59-a78aa2aa4856.png)
I didn't find the Contexts utility class. What is its name?

**Reply by cescoffier** [reply]
Oh, yes, it's `VertxContext` not `Context` 

So, now you just need an executor based on your code you posted:

```runnable -> {
    Context context = VertxContext.getOrCreateDuplicatedContext(vertx);
        VertxContextSafetyToggle.setContextSafe(context, Boolean.TRUE);
        context.runOnContext(x -> {
            runnable.run();
        });
}
```

**Reply by sebastienblanc** [reply]
Funny enough I have been facing the same issue today, and the only way for me to get it working was (method called by onMessage) : 
```
 public Uni<Message> persistMessage(Message message) {
        Context context = vertx.getOrCreateContext();
        Context duplicatedContext = VertxContext.getOrCreateDuplicatedContext(context);
        VertxContextSafetyToggle.setContextSafe(duplicatedContext, Boolean.TRUE);
        duplicatedContext.runOnContext(x -> {
            Panache.withTransaction(()-> message.persist());
        });
        return Uni.createFrom().nothing();
    }
```

But sounds a bit weird 🤣

**Reply by sebastienblanc** [reply]
Well finally this is not working, no error but nothing is written to the DB, even with a `persistAndFlush` , any idea @cescoffier ?

**Reply by cescoffier** [reply]
It looks like a subscription is missing.

**Reply by sebastienblanc** [reply]
Oh Put*** like they would say in French 🤣 
Thanks Clément, indeed I did not subscribe to the uni returned by the `withTransaction` .... 
So this works : 
```
 duplicatedContext.runOnContext(x -> {
            Panache.withTransaction(message::persist).subscribe().with(panacheEntityBase -> Log.info("Persisted "));
        });
```
